lib/versionbot.ts

```diff
@@ -0,0 +1,479 @@
+    // in turn will feed into the 'generateVersion' method below.
+    protected mergePR = (action: GithubBot.GithubAction, data:GithubBot.PullRequestEvent | GithubBot.PullRequestReviewEvent) => {
+        // States for review comments are:
+        //  * COMMENT
+        //  * CHANGES_REQUESTED
+        //  * APPROVED
+        //
+        // We *only* go through with a merge should:
+        //  * The 'flow/ready-to-merge' label appear on the PR issue
+        //  * There is an 'APPROVED' review comment *and* no comment after is of state 'CHANGES_REQUESTED'
+        // The latter overrides the label should it exist, as it will be assumed it is in error.
+        const githubApi = this.githubApi;
+        const pr = data.pull_request;
+        const head = data.pull_request.head;
+        const owner = head.repo.owner.login;
+        const repo = head.repo.name;
+        const repoFullName = `${owner}/${repo}`;
+        const cwd = process.cwd();
+        let newVersion: string;
+        let fullPath: string;
+        let branchName: string;
+        let newTreeSha: string;
+
+        this.log(ProcBot.LogLevel.DEBUG, `${action.name}: entered`);
+
+        // Check the action on the event to see what we're dealing with.
+        switch (data.action) {
+            // Submission is a PR review
+            case 'submitted':
+            case 'labeled':
+            case 'unlabeled':
+                break;
+
+            default:
+                // We have no idea what sparked this, but we're not doing anything!
+                this.log(ProcBot.LogLevel.INFO, `${action.name}:${data.action} isn't a useful action`);
+                return Promise.resolve();
+        }
+
+        // Get the reviews for the PR.
+        return this.gitCall(githubApi.pullRequests.getReviews, {
+            owner: owner,
+            repo: repo,
+            number: pr.number
+        }).then((reviews: GithubBot.Review[]) => {
+            // Cycle through reviews, ensure that any approved review occurred after any requiring changes.
+            let approved = false;
+            // Ooo, this is a good one. We can't do this:
+            //    reviews.forEach((review: GithubBot.Review) => {
+            //      if (review.state === 'APPROVED') {
+            //            approved = true;
+            //      } else if (review.state === 'CHANGES_REQUESTED') {
+            //            approved = false;
+            //      }
+            //     });
+            //
+            // Because of the CFA going on here: https://github.com/Microsoft/TypeScript/issues/11498
+            // This errors because it's optimistic that [].forEach() execution can't affect the
+            // 'approved' variable, therefore it implicitly believes it can only ever be 'false'
+            // and barfs on the 'if (approved === true)' line claiming 'true' cannot be 'false'.
+            // Not great. Still, easy to use a straight for() loop.
+            approved = false;
```

Comment:
You've got `approved = false` here twice.

Also, I think you can get out of the issues you explain in the comment above if you assert `approved` as boolean in the if condition. Not sure if it's significantly better, but an interesting option:

```typescript
if (approved === true) {
  // Doesn't compile, === can't be applied to false and true
}

if (<boolean> approved === true) {
  // Works fine
}
```